src/go/mongolib/profiler/profiler.go

```diff
@@ -217,163 +140,9 @@ func (p *Profile) getDocs() {
 			continue
 		}
 		if len(doc.Query) > 0 {
-			p.ProcessDoc(doc, p.queriesInfoAndCounters)
+			p.stats.Add(doc)
 		}
 	}
-	p.queriesChan <- mapToArray(p.queriesInfoAndCounters)
+	p.queriesChan <- p.stats.Queries()
 	p.Stop()
 }
-
-func (p *Profile) ProcessDoc(doc proto.SystemProfile, stats map[StatsGroupKey]*QueryInfoAndCounters) error {
-
-	fp, err := p.fingerprinter.Fingerprint(doc.Query)
-	if err != nil {
-		return fmt.Errorf("cannot get fingerprint: %s", err.Error())
-	}
-	var s *QueryInfoAndCounters
-	var ok bool
-	p.countersMapLock.Lock()
-	defer p.countersMapLock.Unlock()
-
-	key := StatsGroupKey{
-		Operation:   doc.Op,
-		Fingerprint: fp,
-		Namespace:   doc.Ns,
-	}
-	if s, ok = stats[key]; !ok {
-		realQuery, _ := util.GetQueryField(doc.Query)
-		s = &QueryInfoAndCounters{
-			ID:          fmt.Sprintf("%x", md5.Sum([]byte(fmt.Sprintf("%s", key)))),
-			Operation:   doc.Op,
-			Fingerprint: fp,
-			Namespace:   doc.Ns,
-			TableScan:   false,
-			Query:       realQuery,
-		}
-		stats[key] = s
-	}
-	s.Count++
-	s.NScanned = append(s.NScanned, float64(doc.DocsExamined))
-	s.NReturned = append(s.NReturned, float64(doc.Nreturned))
-	s.QueryTime = append(s.QueryTime, float64(doc.Millis))
-	s.ResponseLength = append(s.ResponseLength, float64(doc.ResponseLength))
-	var zeroTime time.Time
-	if s.FirstSeen == zeroTime || s.FirstSeen.After(doc.Ts) {
-		s.FirstSeen = doc.Ts
-	}
-	if s.LastSeen == zeroTime || s.LastSeen.Before(doc.Ts) {
-		s.LastSeen = doc.Ts
-	}
-
-	return nil
-
-}
-
-func CalcQueriesStats(queries []QueryInfoAndCounters, uptime int64) []QueryStats {
-	stats := []QueryStats{}
-	tc := calcTotalCounters(queries)
-
-	for _, query := range queries {
```

Comment:
Because we need to show them in order for pt-mongodb-query-digest. They are ranked from most used to less used, etc. I think we can de-couple this.